website/docs/configuration/provider-requirements.html.md

```diff
@@ -0,0 +1,299 @@
+unique resource names.
+
+Once you've selected one or more providers, use a `required_providers` block to
+declare them so that Terraform will make them available for use. A provider
+dependency consists of both a source location and a version constraint:
+
+```hcl
+terraform {
+  required_providers {
+    mycloud = {
+      source  = "mycorp/mycloud"
+      version = "~> 1.0"
+    }
+  }
+}
+```
+
+The `required_providers` block must be nested inside a
+[`terraform` block](terraform.html). The `terraform` block can include other
+settings too, but we'll only focus on `required_providers` here.
+
+The keys inside the `required_providers` block represent each provider's
+[local name](#local-names), which is the unique identifier for a provider within
+a particular module. Each item inside the `required_providers` block is an
+object expecting the following arguments:
+
+* `source` - the global [source address](#source-addresses) for the
+  provider you intend to use, such as `hashicorp/aws`.
+
+* `version` - a [version constraint](#version-constraints) specifying
+  which subset of available provider versions the module is compatible with.
+
+-> **Note:** The `required_providers` object syntax described above was added in Terraform v0.13. Previous versions of Terraform used a single string instead of an object, with the string specifying only a version constraint. For example, `mycloud = "~> 1.0"`. Explicit provider source addresses are supported only in Terraform v0.13 and later.
+
+### Source Addresses
+
+A provider _source address_ both globally identifies a particular provider and
+specifies the primary location from which Terraform can download it.
+Source addresses consist of three parts delimited by slashes (`/`), as
+follows:
+
+* **Hostname**: the hostname of the Terraform registry that indexes the provider.
+  You can omit the hostname portion and its following slash if the provider
+  is hosted on [the public Terraform Registry](https://registry.terraform.io/),
+  whose hostname is `registry.terraform.io`.
+
+* **Namespace**: an organizational namespace within the specified registry.
+  For the public Terraform Registry and Terraform Cloud's private registry,
+  this represents the organization that is publishing the provider. This field
+  may have other meanings for other registry hosts.
+
+* **Type**: The provider type name, which must be unique within a particular
+  namespace on a particular registry host.
+
+For example,
+[the official HTTP provider](https://registry.terraform.io/providers/hashicorp/http)
+belongs to the `hashicorp` namespace on `registry.terraform.io`, so its
+source address can be written as either `registry.terraform.io/hashicorp/http`
+or, more commonly, just `hashicorp/http`.
+
+-> **Note**: As a concession for backward compatibility with earlier versions of
+Terraform, the `source` argument is actually optional. If you omit it, Terraform
```

Comment:
I think it's important to specify here (or here-ish) that source is optional _for providers in the hashicorp namespace_? Some providers will not install without source: those in `terraform-providers` which are not moving to the `hashicorp` namespace.

For e.g., `datadog` will not be found if a source isn't specified.

Reply:
The remainder of this paragraph was intending to communicate that very thing, albeit in a different way: rather than saying "it's optional for providers in the hashicorp namespace" it says "if you omit it Terraform will assume the hashicorp namespace". That felt to me like a better way to present the message "you can omit it but we recommend you don't". If you think it's unclear though, I can certainly revise it.

Reply:
Re-reading what you wrote here, I'm seeing that you are probably thinking of the person who is revising their existing configuration that might already be referring to e.g. `datadog`, whereas the audience I had in mind for this was a person who has never used Terraform before and is learning it for the first time here.

The text I drafted in hashicorp/terraform#25113 (the upgrade guide) is intended to say the same thing with the "I have an existing configuration and I'm upgrading" audience in mind. Specifically, it says this at the time I'm writing this comment:

> In order to establish the heirarchical namespace, Terraform now requires explicit source information for any providers that are not HashiCorp-maintained, using a new syntax in the `required_providers` nested block inside the `terraform` configuration block:

What it does _not_ currently say is specifically that "HashiCorp-maintained" means "in the `hashicorp` namespace", so I'd like to suggest that we resolve this by adding some clarification to the upgrade guide text, rather than by changing the prose in the main documentation. Does that compromise seem reasonable to you? :thinking: